src/unit_test.c

```diff
@@ -0,0 +1,407 @@
+    for (i.group = 0; i.group < tf->num_modules; i.group++) {
+        struct TestModule* module = &tf->registry_modules[i.group];
+        int add_all = strcmp(value, module->name) == 0; /* select all from module */
+        for (i.idx = 0; i.idx < module->size; i.idx++) {
+            if (add_all || strcmp(value, module->data[i.idx].name) == 0) {
+                tf->args.targets.slots[tf->args.targets.size] = i;
+                tf->args.targets.size++;
+                /* Matched a single test, we're done */
+                if (!add_all) return 0;
+            }
+        }
+        /* If add_all was true, we added all tests in the module, so return */
+        if (add_all) return 0;
+    }
+    fprintf(stderr, "Error: target '%s' not found (missing or module disabled).\n"
+            "Run program with -print_tests option to display available tests and modules.\n", value);
+    return -1;
+}
+
+/* Read args; all must be "-key=value" */
+static int read_args(int argc, char** argv, int start, struct TestFramework* tf) {
+    int i;
+    char* eq;
+    for (i = start; i < argc; i++) {
+        const char* arg = argv[i];
+        if (!arg || arg[0] != '-') {
+            fprintf(stderr, "Arg '%s' must start with '-'\n", arg ? arg : "(null)");
+            return -1;
+        }
+
+        eq = strchr(arg, '=');
+        if (eq == NULL || eq == arg+1) {
+            fprintf(stderr, "Arg %s must be -key=value\n", arg);
+            return -1;
+        }
+
+        *eq = '\0';
+        if (parse_arg(arg + 1, eq + 1, tf) != 0) {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+static void run_test_log(const struct TestEntry* t) {
+    int64_t start_time = gettime_i64();
+    printf("Running %s..\n", t->name);
+    t->func();
+    printf("Test %s PASSED (%.3f sec)\n", t->name, (double)(gettime_i64() - start_time) / 1000000);
+}
+
+static void run_test(const struct TestEntry* t) { t->func(); }
+
+/* Process tests in sequential order */
+static int run_sequential(struct TestFramework* tf) {
+    int it;
+    for (it = 0; it < tf->args.targets.size; it++) {
+        TestRef* index = &tf->args.targets.slots[it];
+        tf->fn_run_test(&tf->registry_modules[index->group].data[index->idx]);
+    }
+    return EXIT_SUCCESS;
+}
+
+#if SUPPORTS_CONCURRENCY
+/* Process tests in parallel */
+static int run_concurrent(struct TestFramework* tf) {
+    /* Sub-processes info */
+    pid_t workers[MAX_SUBPROCESSES];
+    int pipes[MAX_SUBPROCESSES][2];
+    /* Next worker to send work */
+    int worker_idx;
+    /* Parent process exit status */
+    int status;
+    /* Loop iterator */
+    int it;
+    /* Loop ref */
+    TestRef* ref;
+    /* Launch worker processes */
+    for (it = 0; it < tf->args.num_processes; it++) {
+        pid_t pid;
+        if (pipe(pipes[it]) != 0) {
+            perror("Error during pipe setup");
+            return EXIT_FAILURE;
+        }
+
+        pid = fork();
+        if (pid < 0) {
```

Comment:
318460c39aec04aa3580df5eadc7a935355968de

nit: It would be good to check return values of the `pipe()` and `fork()` calls consistently, for example `...==-1`.